conda_build/variants.py

```diff
@@ -120,195 +137,213 @@ def parse_config_file(path, config):
-                                    "field must be the same length.  In {}, {} and {} are "
-                                    "different ({} and {})".format(spec_source, k, group_item,
-                                                                len(ensure_list(v)),
-                                                                len(ensure_list(spec[group_item]))))
-                                    values[group_item] = ensure_list(spec[group_item])
-                            elif k in values:
-                                for group_item in keys_in_group:
-                                    if group_item in spec and \
-                                            len(ensure_list(spec[group_item])) != len(ensure_list(v)):
-                                        break
-                                    if group_item in values and \
-                                            len(ensure_list(values[group_item])) != len(ensure_list(v)):
-                                        break
-                                else:
-                                    values[k] = v.copy()
-                                if any(subvalue not in values[k]
-                                                    for subvalue in ensure_list(v)):
-
-                                    raise ValueError("variant config in {} is ambiguous because it "
-                                        "does not fully implement all zipped keys, or specifies "
-                                        "a subspace that is not fully implemented.".format(
-                                            spec_source))
-
-    return values
+    errors = []
+    combined = {}
 
+    # check for invalid characters
+    errors.extend(
+        "  - (src:{}) {!r} key contains an invalid character '-'".format(src, k)
+        for src, spec in specs.items()
+        for k in spec
+        if "-" in k
+    )
 
-def combine_specs(specs, log_output=True):
-    """With arbitrary sets of sources, combine into a single aggregate spec.
+    # extend zip_keys while checking to see if any keys are duplicated
+    zip_keys = {}
+    seen = {}
+    for src, spec in specs.items():
+        try:
+            zk = _get_zip_keys(spec)
+        except ValueError as e:
+            errors.append("  - (src:{}) {}".format(src, e))
+        else:
+            errors.extend(
+                "  - key {!r} has been used multiple times in zip_keys, keys can only occur in one group:\n"
+                "      (src:{}) {}\n"
+                "      (src:{}) {}".format(k, seen[k]["src"], tuple(seen[k]["zg"]), src, tuple(zg))
+                if k in seen else seen.setdefault(k, {}).update({"src": src, "zg": zg})
+                for zg in zk
+                if zg not in zip_keys
+                for k in zg
+            )
+            zip_keys.update({zg: src for zg in zk})
+    # only include zip_keys if they are explicitly defined
+    if zip_keys:
+        combined["zip_keys"] = list(zip_keys)
```

Comment:
I personally do not see any danger in passing through an empty zip_keys and would rather not have this if clause.

I added this check here to pass `tests/test_variants::test_later_spec_priority`, the same is done for extend_keys below.